src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,180 @@
             }
-
-            return 0;
+            await serializer.SaveAsync(solutionFileFullPath, solution, cancellationToken);
         }
 
-        private static IList<string> GetSolutionFoldersFromProjectPath(string projectFilePath)
+        private void AddProject(SolutionModel solution, string pathRelativeToSolutionFile, string fullPath, SolutionFolderModel solutionFolder)
         {
-            var solutionFolders = new List<string>();
-
-            if (!IsPathInTreeRootedAtSolutionDirectory(projectFilePath))
-                return solutionFolders;
-
-            var currentDirString = $".{Path.DirectorySeparatorChar}";
-            if (projectFilePath.StartsWith(currentDirString))
+            // Open project instance to see if it is a valid project
+            ProjectRootElement projectRootElement = ProjectRootElement.Open(fullPath);
+            SolutionProjectModel project;
+            try
             {
-                projectFilePath = projectFilePath.Substring(currentDirString.Length);
+                project = solution.AddProject(pathRelativeToSolutionFile, null, solutionFolder);
             }
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
-        }
-
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
-        {
-            if (_inRoot)
+            catch (SolutionArgumentException ex) when (ex.ParamName == "projectTypeName")
             {
-                // The user requested all projects go to the root folder
-                return null;
+                // If guid is not identified by vs-solutionpersistence, check in project element itself
+                var guid = projectRootElement.GetProjectTypeGuid();
+                if (string.IsNullOrEmpty(guid))
+                {
+                    Reporter.Error.WriteLine(CommonLocalizableStrings.UnsupportedProjectType, fullPath);
+                    return;
+                }
+                project = solution.AddProject(pathRelativeToSolutionFile, guid, solutionFolder);
             }
-
-            if (_relativeRootSolutionFolders != null)
+            // Generate intermediate solution folders
+            if (solutionFolder is null && !_inRoot)
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
```

Comment:
In the old behavior, you could do dotnet sln add -s ..\foo r. It would add .. to the folder as a dir and then put r/ in r/r.csproj if there was a duplicate project not called r in the folder foo. Now, the API fails: One or more errors occurred. (Names cannot:

contain any of the following characters: / ? : \ * " < > |
contain control characters
be system reserved names, including 'CON', 'AUX', 'PRN', 'COM1' or 'LPT2'
be '.' or '..' (Parameter 'name'))
Is this an expected breaking change?

Reply:
The new API does have a method that validates names for both projects and solution folders. The restrictions make sense to me considering OS' rules, but if it did work on the old behavior, it is possible this _is_ a breaking change

Reply:
Ok, awesome. That's not a blocker (I mean its not even within our control) but we should mark this as a breaking change and it likely needs a doc, if that hasn't been said already. I don't know if youre keeping a list of stuff but I would recommend you do.